python/traverse.py

```diff
@@ -26,50 +26,70 @@
 }
 
 
-def traverse_and_execute(
-    html: str, fn: Callable[[lxml.html.HtmlElement], None]
-) -> None:
-    elements = lxml.html.fragments_fromstring(html)
+def get_elements_list(html: str) -> List[slp.Node]:
+    parser = slp.HTMLParser(html)
+    res = []
 
-    # If there's leading text, the first element of the array will be a string.
-    # We can just discard that.
-    if isinstance(elements[0], str):
-        del elements[0]
+    if parser.head is not None:
+        res.extend(parser.head.iter(include_text=True))
 
-    for element in elements:
-        for e in element.iter():
-            fn(e)
+    if parser.body is not None:
+        res.extend(parser.body.iter(include_text=True))
 
+    return res
 
-def format_attrib_value(v: str) -> str:
+
+def traverse_and_execute(html: str, fn: Callable[[slp.Node], None]) -> None:
+    # elements = get_elements_list(html)
+
+    page = slp.HTMLParser(html)
+
+    if page.head is None:
```

Comment:
Does `head` correspond to `<head>`, or something else? If the former, why are we iterating over the nodes in head?

Reply:
Discussed in https://github.com/PrairieLearn/PrairieLearn/pull/7179/files/1cbd286caa0eec6a789665879de5a1d186dba96b#r1117968904. Basically, the parser treats `html` as a complete page and automatically creates `head` and `body` tags. In this case, the body is treated as one of the children of the head tag, so calling `traverse` in head goes through all the nodes in the page.